src/internal/packager/images/pull.go

```diff
@@ -99,52 +150,291 @@ func (i *ImgConfig) PullAll() error {
+		cranePath, err = layout.Write(i.ImagesPath, empty.Index)
+		if err != nil {
+			return err
+		}
+	}
+
+	for tag, img := range tagToImage {
+		imgDigest, err := img.Digest()
+		if err != nil {
+			return fmt.Errorf("unable to get digest for image %s: %w", tag, err)
+		}
+		tagToDigest[tag.String()] = imgDigest.String()
+	}
+
 	spinner.Success()
 
 	// Create a thread to update a progress bar as we save the image files to disk
 	doneSaving := make(chan int)
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go utils.RenderProgressBarForLocalDirWrite(i.ImagesPath, totalBytes, &wg, doneSaving, fmt.Sprintf("Pulling %d images", imgCount))
+	var progressBarWaitGroup sync.WaitGroup
+	progressBarWaitGroup.Add(1)
+	go utils.RenderProgressBarForLocalDirWrite(i.ImagesPath, totalBytes, &progressBarWaitGroup, doneSaving, fmt.Sprintf("Pulling %d images", imgCount))
+
+	// Spawn a goroutine for each layer to write it to disk using crane
+
+	layerWritingConcurrency := utils.NewConcurrencyTools[bool, error](len(processedLayers))
+
+	defer layerWritingConcurrency.Cancel()
+
+	for _, layer := range processedLayers {
+		layer := layer
+		// Function is a combination of https://github.com/google/go-containerregistry/blob/v0.15.2/pkg/v1/layout/write.go#L270-L305
+		// and https://github.com/google/go-containerregistry/blob/v0.15.2/pkg/v1/layout/write.go#L198-L262
+		// with modifications. This allows us to dedupe layers for all images and write them concurrently.
+		go func() {
+			defer layerWritingConcurrency.WaitGroupDone()
+			digest, err := layer.Digest()
+			if errors.Is(err, stream.ErrNotComputed) {
+				// Allow digest errors, since streams may not have calculated the hash
+				// yet. Instead, use an empty value, which will be transformed into a
+				// random file name with `os.CreateTemp` and the final digest will be
+				// calculated after writing to a temp file and before renaming to the
+				// final path.
+				digest = v1.Hash{Algorithm: "sha256", Hex: ""}
+			} else if err != nil {
+				layerWritingConcurrency.ErrorChan <- err
+				return
+			}
+
+			size, err := layer.Size()
+			if errors.Is(err, stream.ErrNotComputed) {
+				// Allow size errors, since streams may not have calculated the size
+				// yet. Instead, use -1 as a sentinel value meaning that no size
+				// comparison can be done and any sized blob file should be considered
+				// valid and not overwritten.
+				//
+				// TODO: Provide an option to always overwrite blobs.
+				size = -1
+			} else if err != nil {
+				layerWritingConcurrency.ErrorChan <- err
+				return
+			}
+
+			if layerWritingConcurrency.IsDone() {
+				return
+			}
+
+			readCloser, err := layer.Compressed()
+			if err != nil {
+				layerWritingConcurrency.ErrorChan <- err
+				return
+			}
+
+			// Create the directory for the blob if it doesn't exist
+			dir := filepath.Join(string(cranePath), "blobs", digest.Algorithm)
+			if err := os.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
```

Comment:
```suggestion
			if err := utils.CreateDirectory(dir, os.ModePerm); err != nil {
```
Per @Noxsios 's comment (he may have selected the wrong line)

Reply:
changed